engine/verification/assigner/engine.go

```diff
@@ -1,30 +1,39 @@
 package assigner
 
 import (
+	"context"
+	"fmt"
+
+	"github.com/opentracing/opentracing-go"
 	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
 
 	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/chunks"
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/trace"
 	"github.com/onflow/flow-go/state/protocol"
 	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
 )
 
 // The Assigner engine reads the receipts from each finalized block.
 // For each receipt, it reads its result and find the chunks the assigned
 // to me to verify, and then save it to the chunks job queue for the
 // fetcher engine to process.
 type Engine struct {
-	unit    *engine.Unit
-	log     zerolog.Logger
-	metrics module.VerificationMetrics
-	me      module.Local
-	state   protocol.State
-
-	assigner         module.ChunkAssigner  // used to determine chunks this node needs to verify
-	chunksQueue      storage.ChunksQueue   // to store chunks to be verified
-	newChunkListener module.NewJobListener // to notify about a new chunk
-	finishProcessing finishProcessing      // to report a block has been processed
+	unit                  *engine.Unit
+	log                   zerolog.Logger
+	metrics               module.VerificationMetrics
+	tracer                module.Tracer
+	me                    module.Local
+	state                 protocol.State
+	assigner              module.ChunkAssigner  // to determine chunks this node should verify.
+	chunksQueue           storage.ChunksQueue   // to store chunks to be verified.
+	newChunkListener      module.NewJobListener // to notify chunk queue consumer about a new chunk.
+	blockConsumerNotifier ProcessingNotifier    // to report a block has been processed.
```

Comment:
I understand `newChunkListener` -- the way the queue works, we need to ping its goroutine when we insert a job to storage. What is the `blockConsumerNotifier` for?

Reply:
The `assigner` engine receives blocks from a block consumer. The block consumer is acting as a faucet that controls the flow of blocks to the `assigner` engine. The flow in the block consumer faucet is controlled with this `blockConsumerNotifier`. Once `assigner` is (really!) done with a block, it notifies the block consumer to push the next finalized block. Here, by `assigner` is being _really done_ with a block, we mean it processes all its receipts and returns on it _gracefully_ and does not for example return of a failure. Also, the block consumer fires an `assigner` worker per each block (concurrently) while limiting the max number of workers. So, letting block consumer notified of a graceful termination, it can fire up the next block on the free slot of an available `assigner` worker.

Reply:
Definitely out of scope for this PR, but just noting it might make sense to replace this interaction with message queues between engines as described in https://github.com/onflow/flow/issues/396.